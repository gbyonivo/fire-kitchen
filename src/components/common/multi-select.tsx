import {
  Label,
  Listbox,
  ListboxButton,
  ListboxOption,
  ListboxOptions,
} from "@headlessui/react";
import { CheckIcon, ChevronDownIcon } from "@heroicons/react/24/solid";
import clsx from "clsx";
import { TextInput } from "./text-input";
import { useMemo, useState } from "react";

export interface MultiSelectProps<T> {
  options: T[];
  selectedOptions: T[];
  onChange: (options: T[]) => void;
  getLabel: (option: T) => string;
  getKey: (option: T) => string;
  getSearchKey: (option: T) => string;
  label?: string;
  containerClassName?: string;
}

export function MultiSelect<T>({
  options,
  selectedOptions,
  onChange,
  getKey,
  getLabel,
  getSearchKey,
  label,
  containerClassName,
}: MultiSelectProps<T>) {
  const [search, setSearch] = useState("");
  const displayedOptions = useMemo(() => {
    return options.filter((option) =>
      getSearchKey(option).toLowerCase().includes(search.toLowerCase())
    );
  }, [options, getSearchKey, search]);

  return (
    <div className={containerClassName}>
      <Listbox value={selectedOptions} onChange={onChange} multiple>
        {label && (
          <Label className="block text-sm/6 font-medium text-gray-400">
            {label}
          </Label>
        )}
        <ListboxButton
          className={clsx(
            "relative block w-full rounded-lg bg-white/5 py-1.5 pr-8 pl-3 text-left text-sm/6 text-white min-h-10",
            "focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
          )}
        >
          <div className="flex flex-wrap gap-2">
            {selectedOptions.map((option) => (
              <div
                key={getKey(option)}
                className="rounded-lg bg-white/10 px-2 py-1 text-sm/6 text-white h-6 text-xs"
                onClick={() => {
                  onChange(selectedOptions.filter((o) => o !== option));
                }}
              >
                {getLabel(option)}
              </div>
            ))}
          </div>
          <ChevronDownIcon
            className="group pointer-events-none absolute top-2.5 right-2.5 size-4 fill-white/60"
            aria-hidden="true"
          />
        </ListboxButton>
        <ListboxOptions
          anchor="bottom"
          transition
          className={clsx(
            "w-(--button-width) rounded-xl border border-white/5 bg-black p-1 [--anchor-gap:--spacing(1)] focus:outline-none",
            "transition duration-100 ease-in data-leave:data-closed:opacity-0"
          )}
        >
          <TextInput value={search} onChange={(value) => setSearch(value)} />
          <div className="h-64 overflow-y-auto py-4">
            {displayedOptions.map((option) => (
              <ListboxOption
                key={getKey(option)}
                value={option}
                className="group flex cursor-default items-center gap-2 rounded-lg px-3 py-1.5 select-none data-focus:bg-white/10"
              >
                <CheckIcon className="invisible size-4 fill-white group-data-selected:visible" />
                <div className="text-sm/6 text-white">{getLabel(option)}</div>
              </ListboxOption>
            ))}
          </div>
        </ListboxOptions>
      </Listbox>
    </div>
  );
}
