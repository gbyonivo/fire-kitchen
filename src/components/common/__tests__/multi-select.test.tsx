import { render } from "@testing-library/react";
import { MultiSelect, MultiSelectProps } from "../multi-select";

const optionsAsObjArray = [
  { id: "1", name: "Option 1" },
  { id: "2", name: "Option 2" },
  { id: "3", name: "Option 3" },
];

const objProps: Partial<MultiSelectProps<{ id: string; name: string }>> = {
  options: optionsAsObjArray,
  getLabel: (option: { id: string; name: string }) => option.name,
  getKey: (option: { id: string; name: string }) => option.id,
  getSearchKey: (option: { id: string; name: string }) => option.name,
};

const optionsAsStrArray = optionsAsObjArray.map((option) => option.name);

const strProps: Partial<MultiSelectProps<string>> = {
  options: optionsAsStrArray,
  getLabel: (option: string) => option,
  getKey: (option: string) => option,
  getSearchKey: (option: string) => option,
};

describe("MultiSelect", () => {
  test.each([[strProps], [objProps]])("%s should match snapshot", (props) => {
    const onChange = jest.fn();
    const { container } = render(
      // @ts-expect-error - test
      <MultiSelect selectedOptions={[]} onChange={onChange} {...props} />
    );
    expect(container).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
