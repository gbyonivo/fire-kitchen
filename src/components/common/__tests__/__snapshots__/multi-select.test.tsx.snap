// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect {
  options: [Array],
  getLabel: [Function: getLabel],
  getKey: [Function: getKey],
  getSearchKey: [Function: getSearchKey]
} should match snapshot 1`] = `
<div>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="relative block w-full rounded-lg bg-white/5 py-1.5 pr-8 pl-3 text-left text-sm/6 text-white min-h-10 focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
      data-headlessui-state=""
      id="headlessui-listbox-button-«r2»"
      type="button"
    >
      <div
        class="flex flex-wrap gap-2"
      />
      <svg
        aria-hidden="true"
        class="group pointer-events-none absolute top-2.5 right-2.5 size-4 fill-white/60"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.53 16.28a.75.75 0 0 1-1.06 0l-7.5-7.5a.75.75 0 0 1 1.06-1.06L12 14.69l6.97-6.97a.75.75 0 1 1 1.06 1.06l-7.5 7.5Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`MultiSelect {
  options: [Array],
  getLabel: [Function: getLabel],
  getKey: [Function: getKey],
  getSearchKey: [Function: getSearchKey]
} should match snapshot 2`] = `
<div>
  <div>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="relative block w-full rounded-lg bg-white/5 py-1.5 pr-8 pl-3 text-left text-sm/6 text-white min-h-10 focus:not-data-focus:outline-none data-focus:outline-2 data-focus:-outline-offset-2 data-focus:outline-white/25"
      data-headlessui-state=""
      id="headlessui-listbox-button-«r8»"
      type="button"
    >
      <div
        class="flex flex-wrap gap-2"
      />
      <svg
        aria-hidden="true"
        class="group pointer-events-none absolute top-2.5 right-2.5 size-4 fill-white/60"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.53 16.28a.75.75 0 0 1-1.06 0l-7.5-7.5a.75.75 0 0 1 1.06-1.06L12 14.69l6.97-6.97a.75.75 0 1 1 1.06 1.06l-7.5 7.5Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
